crypto/src/merkle_tree/mod.rs

```diff
@@ -0,0 +1,267 @@
+}
+
+impl<F: IsField, H: IsCryptoHash<F> + Clone> MerkleTree<F, H> {
+    pub fn build(values: &[FieldElement<F>], hasher: H) -> MerkleTree<F, H> {
+        let mut leafs: Vec<TreeNode<F>> = hash_leafs(values, &hasher);
+        let mut level: Vec<TreeNode<F>> = hash_level(&mut leafs, &hasher);
+
+        while level.len() > 1 {
+            level = hash_level(&mut level, &hasher);
+        }
+
+        MerkleTree {
+            leafs,
+            root: level[0].clone(),
+            hasher,
+        }
+    }
+
+    pub fn get_root_hash(&self) -> FieldElement<F> {
+        self.root.borrow().hash.clone()
+    }
+
+    pub fn get_proof(&self, value: FieldElement<F>) -> Option<Proof<F, H>> {
+        let hashed_value = self.hasher.hash_one(value.clone());
+
+        if let Some(leaf) = self
+            .leafs
+            .iter()
+            .find(|node| node.borrow().hash == hashed_value)
+        {
+            let mut merkle_path: Vec<TreeNode<F>> = Vec::new();
+            merkle_path = build_merkle_path(leaf.clone(), &mut merkle_path).to_vec();
+
+            return Some(Proof {
+                value,
+                merkle_path,
+                hasher: self.hasher.clone(),
+            });
+        }
+
+        None
+    }
+
+    pub fn verify(proof: &Proof<F, H>, root_hash: FieldElement<F>) -> bool {
+        let mut hashed_value = proof.hasher.hash_one(proof.value.clone());
+
+        for node in proof.merkle_path.iter() {
+            if let Some(sibiling) = &node.borrow().sibiling {
+                hashed_value = proof
+                    .hasher
+                    .hash_two(hashed_value, sibiling.borrow().hash.clone());
+            }
+        }
+        hashed_value == root_hash
+    }
+}
+
+fn hash_leafs<F: IsField, H: IsCryptoHash<F>>(
+    values: &[FieldElement<F>],
+    hasher: &H,
+) -> Vec<TreeNode<F>> {
+    values
+        .iter()
+        .map(|value| build_tree_node(hasher.hash_one(value.clone())))
+        .collect()
+}
+
+fn hash_level<F: IsField, H: IsCryptoHash<F>>(
+    values: &mut Vec<TreeNode<F>>,
+    hasher: &H,
+) -> Vec<TreeNode<F>> {
+    if values.len() % 2 != 0 {
+        values.push(values[values.len() - 1].clone());
+    }
+
+    values
+        .chunks(2)
+        .map(|chunk| build_parent(chunk[0].clone(), chunk[1].clone(), hasher))
+        .collect()
+}
+
+fn build_parent<F: IsField, H: IsCryptoHash<F>>(
+    left: TreeNode<F>,
+    right: TreeNode<F>,
+    hasher: &H,
+) -> TreeNode<F> {
+    let parent_hash = hasher.hash_two(left.borrow().hash.clone(), right.borrow().hash.clone());
+    let parent = build_tree_node(parent_hash);
+
+    left.borrow_mut().sibiling = Some(right.clone());
+    left.borrow_mut().parent = Some(parent.clone());
+
+    right.borrow_mut().sibiling = Some(left);
+    right.borrow_mut().parent = Some(parent.clone());
+
+    parent
+}
+
+fn build_merkle_path<F: IsField>(
+    node: TreeNode<F>,
+    merkle_path: &mut Vec<TreeNode<F>>,
+) -> &Vec<TreeNode<F>> {
```

Comment:
Could we return a `Vec<TreeNode<F>>` instead of its reference? With this change you remove the need for the caller of this function to use the `to_vec()` method.